File: bindings/react-native-pirate-wallet/ios/PirateWalletReactNative.m
#import <React/RCTBridgeModule.h>
@import PirateWalletNative;

@interface PirateWalletReactNative : NSObject <RCTBridgeModule>
@end

@implementation PirateWalletReactNative

RCT_EXPORT_MODULE();

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

RCT_REMAP_METHOD(invoke,
                 invoke:(NSString *)requestJson
                 pretty:(BOOL)pretty
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  const char *requestCString = [requestJson UTF8String];
  if (requestCString == NULL) {
    reject(@"PIRATE_WALLET_INVOKE_ERROR", @"Request string was not valid UTF-8.", nil);
    return;
  }

  char *responsePtr = pirate_wallet_service_invoke_json(requestCString, pretty);
  if (responsePtr == NULL) {
    reject(@"PIRATE_WALLET_INVOKE_ERROR", @"Wallet service returned a null response.", nil);
    return;
  }

  NSString *response = [NSString stringWithUTF8String:responsePtr];
  pirate_wallet_service_free_string(responsePtr);

  if (response == nil) {
    reject(@"PIRATE_WALLET_INVOKE_ERROR", @"Wallet service returned invalid UTF-8.", nil);
    return;
  }

  resolve(response);
}

@end

File: bindings/react-native-pirate-wallet/ios/PirateWalletReactNative.swift
import Foundation
import PirateWalletNative
import React

@objc(PirateWalletReactNative)
final class PirateWalletReactNative: NSObject {
  @objc
  static func requiresMainQueueSetup() -> Bool {
    false
  }

  @objc
  func invoke(
    _ requestJson: String,
    pretty: Bool,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let request = requestJson.cString(using: .utf8) else {
      reject("PIRATE_WALLET_INVOKE_ERROR", "Request string was not valid UTF-8.", nil)
      return
    }

    let pointer = request.withUnsafeBufferPointer { buffer in
      pirate_wallet_service_invoke_json(buffer.baseAddress, pretty)
    }

    guard let pointer else {
      reject("PIRATE_WALLET_INVOKE_ERROR", "Wallet service returned a null response.", nil)
      return
    }

    defer {
      pirate_wallet_service_free_string(pointer)
    }

    let response = String(cString: pointer)
    resolve(response)
  }
}
